tests/hmm/hmumu_utils.py

```diff
@@ -1520,6 +1527,61 @@ def get_selected_jets(
 
     return ret
 
+
+def get_puid_weights(jets, passed_puid, extractors, hist_names):
+
+    nev = jets.numevents()
+
+    jets_pu_eff, jets_pu_sf = jet_puid_evaluate(extractors, hist_names, jets.pt, jets.eta)
+    p_puid_mc = NUMPY_LIB.zeros(nev)
+    p_puid_data = NUMPY_LIB.zeros(nev)
+    compute_eff_product(jets.offsets, passed_puid, jets_pu_eff, p_puid_mc)
+    compute_eff_product(jets.offsets, passed_puid, jets_pu_eff*jets_pu_sf, p_puid_data)
+    eventweight_puid = np.divide(p_puid_data, p_puid_mc, out=np.zeros_like(p_puid_data), where=p_puid_mc!=0)
+
+    return eventweight_puid
+
+def jet_puid_extract(calibration_file, h_eff_name, h_sf_name):
+    print("Extracting PUID eff. from "+h_eff_name)
+    ext_eff = extractor()
+    ext_eff.add_weight_sets(["{0} {0} {1}".format(h_eff_name, calibration_file)])
+
+    print("Extracting PUID SF from "+h_sf_name)
+    ext_sf = extractor()
+    ext_sf.add_weight_sets(["{0} {0} {1}".format(h_sf_name, calibration_file)])
+
+    extractors = (ext_eff, ext_sf)
+    hist_names = (h_eff_name, h_sf_name)
+    return extractors, hist_names
+
+def jet_puid_evaluate(extractors, hist_names, jet_pt, jet_eta):
+    ext_eff, ext_sf = extractors
+    h_eff_name, h_sf_name = hist_names
+
+    ext_eff.finalize()
+    evaluator_eff = ext_eff.make_evaluator()
```

Comment:
Very good!

I think you should also be able to move the `finalize()` and `make_evaluator()` to the function `jet_puid_extract()`, such that this function becomes `jet_puid_evalute(puid_evaluator, jet_pt, jet_eta)` and already takes just the object `puid_evaluator=evaluator_eff[h_eff_name]` as one of the arguments.

This way you don't have to redo the finalization every time you wan to compute the puid values.

See e.g. https://github.com/jpata/hepaccelerate-cms/blob/master/tests/hmm/analysis_hmumu.py#L103 how the JES stuff is finalized once in the beginning of the code.